File: 1812/Domain/Player.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace _1812.Domain
{
    public class Player
    {
        public Tablet tablet { get; private set; }
        public List<Card> hand { get; private set; }
        private static List<Card> toDrop { get; set; }
        public Player(Tablet tablet, List<Card> hand)
        {
            this.tablet = tablet;
            this.hand = hand;
        }
        public static List<Card> CleanBuffer()
        {
            List<Card> toReturn = toDrop;
            toDrop = new List<Card>();
            return toReturn;
        }
        public Interaction FirstPhase(int Dice)
        {
            Interaction interaction = new Interaction();

            List<Api.ActionType> allowedActions = new List<Api.ActionType>();

            int cannonCount = 0;
            foreach (Card i in hand)
            {
                if (i.Type == Card.CardType.Cannon)
                {
                    cannonCount++;
                }
            }

            if (cannonCount != 0)
            {
                allowedActions.Add(Api.ActionType.PerformKnockout);
            }
            if (tablet.operation != null)
            {
                allowedActions.Add(Api.ActionType.UseOperation);
            }
            allowedActions.Add(Api.ActionType.AttackCommander);
            allowedActions.Add(Api.ActionType.DamageCamp);

            Api.ActionType action = Api.ChooseAction(allowedActions);

            if (action == Api.ActionType.AttackCommander)
            {
                interaction.Attack = tablet.ActCommanderAttack(Dice);
            }
            if (action == Api.ActionType.DamageCamp)
            {
                interaction.Attack = tablet.ActCampAttack(Dice);
            }
            if (action == Api.ActionType.UseOperation)
            {
                Console.WriteLine("PLAYER FIRST PHASE OPERATION");
            }
            if (action == Api.ActionType.PerformKnockout)
            {
                List<Card.CardType> target = Api.ChooseWhatToKnockout();
                List<Card> weapon = Api.ChooseCardsToKnockoutWith(hand);
                Tuple<int, List<Card.CardType>> knockoutAction = tablet.ActKnockout(weapon, target);

                foreach (Card i in weapon) // убрать карты, которыми атакуем
                {
                    hand.Remove(i);
                }
                interaction.KnockOut = knockoutAction;
            }
            return interaction;
        }
    }
}

File: domain/player.py
from domain.tablet import Tablet
from typing import List
from domain.card import Card, CardType
from domain.interaction import Interaction
from domain.api import Api, ActionType

class Player:

    def __init__(self, tablet: Tablet, hand: List[Card]):
        self.tablet = tablet
        self.hand = hand
        self.to_drop = []

    def clean_buffer(self):
        to_return = self.to_drop.copy()
        self.to_drop = []
        return to_return

    def fill_tablet(self):
        # select action, cannon, 
        self.hand.filter()

    def first_phase(self, dice: int):
        interaction = Interaction()
        allowed_actions = []
        cannon_count = 0
        for i in self.hand:
            if i.type_id is CardType.Cannon:
                cannon_count += 1
        
        if cannon_count != 0:
            allowed_actions.append(ActionType.PerformKnockout)
        
        if self.tablet.operation != None:
            allowed_actions.append(ActionType.UseOperation)
        
        allowed_actions.append(ActionType.AttackCommander)
        allowed_actions.append(ActionType.DamageCamp)

        action = Api.choose_action(allowed_actions)

        if action is ActionType.AttackCommander:
            interaction.Attack = self.tablet.act_commander_attack(dice)

        if action is ActionType.DamageCamp:
            interaction.Attack = self.tablet.act_camp_attack(dice)

        if action is ActionType.UseOperation:
            print("PLAYER FIRST PHASE OPERATION")

        if action is ActionType.PerformKnockout:
            target = Api.choose_what_to_knockout()
            weapon = Api.choose_cards_to_knockout_with(self.hand)
            knockoutAction = self.tablet.act_knockout(weapon, target)

            for i in weapon:
                self.hand.Remove(i)

            interaction.knock_out = knockoutAction
        return interaction

    